src/backend/commands/copy.c

```diff
@@ -472,9 +472,18 @@ DoCopy(ParseState *pstate, const CopyStmt *stmt,
 		 */
 		PG_TRY();
 		{
-			cstate = BeginCopyTo(pstate, rel, query, relid,
-								 stmt->filename, stmt->is_program,
-								 stmt->attlist, options);
+			if (rel && rel->rd_rel->relkind == RELKIND_DIRECTORY_TABLE)
+			{
+				cstate = BeginCopyToDirectoryTable(pstate, stmt->filename, stmt->dirfilename,
+												   rel, stmt->is_program, options);
+			}
+
+			else
+			{
+				cstate = BeginCopyTo(pstate, rel, query, relid,
+									 stmt->filename, stmt->is_program,
+									 stmt->attlist, options);
+			}
```

Comment:
Could rel be NULL? and if so, will it crash in else branch?

Reply:
No, BecinCopyTo can accept rel is NULL and it processed in logic.